"Expected indentation of NaN spaces"

<!--
This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version:** v4.0.0-alpha.2
* **Node Version:** 7.10.0
* **npm Version:** 4.2.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

No configuration file used. `--no-eslintrc` flag passed on CLI.

<!-- Paste your configuration below: -->
```
```

**What did you do? Please include the actual source code causing the issue.**

Command run was: `eslint --no-eslintrc test.js`. Contents of `test.js`:

<!-- Paste the source code below: -->
```js
/* eslint indent: [2, 2, {ArrayExpression: first, CallExpression: {arguments: first}, FunctionDeclaration: {parameters: first}, FunctionExpression: {parameters: first},  MemberExpression: off, ObjectExpression: first, SwitchCase: 1}] */

function rtfEscape(string) {
  return string
    .replace(/[\\{}]/g, function(m) {
      return '\\' + m;
    })
    .replace(/\t/g, function() {
      return '\\tab ';
    })
    // eslint-disable-next-line no-control-regex
    .replace(/[\x00-\x1f\x7f-\xff]/g, function(m) {
      return '\\\'' + toHex(m.charCodeAt(0), 2);
    })
    .replace(/\ufeff/g, '')
    .replace(/[\u0100-\uffff]/g, function(m) {
      return '\\u' + toHex(m.charCodeAt(0), 4) + '?';
    });
}
```

**What did you expect to happen?**

Any indentation expectations would be reported as an integer.

**What actually happened? Please include the actual, raw output from ESLint.**

Indentation expectations reported as `NaN`:

```console
   6:1  error  Expected indentation of NaN spaces but found 6  indent
   7:1  error  Expected indentation of NaN spaces but found 4  indent
   9:1  error  Expected indentation of NaN spaces but found 6  indent
  10:1  error  Expected indentation of NaN spaces but found 4  indent
  13:1  error  Expected indentation of NaN spaces but found 6  indent
  14:1  error  Expected indentation of NaN spaces but found 4  indent
  17:1  error  Expected indentation of NaN spaces but found 6  indent
  18:1  error  Expected indentation of NaN spaces but found 4  indent
```

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -924,21 +924,49 @@ module.exports = {
 
             MemberExpression(node) {
                 const firstNonObjectToken = sourceCode.getFirstTokenBetween(node.object, node.property, astUtils.isNotClosingParenToken);
-                const tokensToIndent = node.computed ? [sourceCode.getTokenAfter(firstNonObjectToken)] : [firstNonObjectToken, sourceCode.getTokenAfter(firstNonObjectToken)];
+                const secondNonObjectToken = sourceCode.getTokenAfter(firstNonObjectToken);
+
+                const tokenBeforeObject = sourceCode.getTokenBefore(node.object, astUtils.isNotOpeningParenToken);
+                const firstObjectToken = tokenBeforeObject ? sourceCode.getTokenAfter(tokenBeforeObject) : sourceCode.ast.tokens[0];
+                const lastObjectToken = sourceCode.getTokenBefore(firstNonObjectToken);
 
                 if (node.computed) {
+
+                    // For computed MemberExpressions, match the closing bracket with the opening bracket.
                     offsets.matchIndentOf(firstNonObjectToken, sourceCode.getLastToken(node));
                 }
 
-                if (typeof options.MemberExpression !== "number") {
-                    tokensToIndent.forEach(token => offsets.ignoreToken(token));
+                if (typeof options.MemberExpression === "number") {
+                    const firstPropertyToken = node.computed ? firstNonObjectToken : secondNonObjectToken;
+
+                    /*
+                     * If the object ends on the same line that the property starts, match against the last token
+                     * of the object, to ensure that the MemberExpression is not indented.
+                     *
+                     * Otherwise, match against the first token of the object, e.g.
+                     * foo
+                     *   .bar
+                     *   .baz // <-- offset by 1 from `foo`
+                     */
+                    const offsetBase = lastObjectToken.loc.end.line === firstPropertyToken.loc.start.line
+                        ? lastObjectToken
+                        : firstObjectToken;
+
+                    // Match the dot (for non-computed properties) or the opening bracket (for computed properties) against the object.
+                    offsets.setDesiredOffset(firstNonObjectToken, offsetBase, options.MemberExpression);
+
+                    /*
+                     * For computed MemberExpressions, match the first token of the property against the opening bracket.
+                     * Otherwise, match the first token of the property against the object.
+                     */
+                    offsets.setDesiredOffset(secondNonObjectToken, node.computed ? firstNonObjectToken : offsetBase, options.MemberExpression);
+                } else {
+
+                    // If the MemberExpression option is off, ignore the dot and the first token of the property.
+                    offsets.ignoreToken(firstNonObjectToken);
+                    offsets.ignoreToken(secondNonObjectToken);
+                    offsets.matchIndentOf(firstNonObjectToken, secondNonObjectToken);
                 }
-
-                const offsetBase = node.object.loc.end.line === node.property.loc.start.line
-                    ? sourceCode.getLastToken(node.object)
-                    : sourceCode.getFirstToken(node.object);
-
-                offsets.setDesiredOffsets(tokensToIndent, offsetBase, options.MemberExpression);
             },
 
             NewExpression(node) {

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -926,7 +926,7 @@ module.exports = {
                 const firstNonObjectToken = sourceCode.getFirstTokenBetween(node.object, node.property, astUtils.isNotClosingParenToken);
                 const secondNonObjectToken = sourceCode.getTokenAfter(firstNonObjectToken);
 
-                const tokenBeforeObject = sourceCode.getTokenBefore(node.object, astUtils.isNotOpeningParenToken);
+                const tokenBeforeObject = sourceCode.getTokenBefore(node.object, token => astUtils.isNotOpeningParenToken(token) || parameterParens.has(token));
                 const firstObjectToken = tokenBeforeObject ? sourceCode.getTokenAfter(tokenBeforeObject) : sourceCode.ast.tokens[0];
                 const lastObjectToken = sourceCode.getTokenBefore(firstNonObjectToken);
 

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -3272,6 +3272,15 @@ ruleTester.run("indent", rule, {
                     baz();
                 })
             `
+        },
+        {
+            code: unIndent`
+                new Foo(
+                    bar
+                        .baz
+                        .qux
+                )
+            `
         }
     ],
 

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -3166,6 +3166,112 @@ ruleTester.run("indent", rule, {
                 }
             `,
             parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: unIndent`
+                foo
+                    .bar(function() {
+                        baz
+                    })
+            `
+        },
+        {
+            code: unIndent`
+                foo
+                        .bar(function() {
+                            baz
+                        })
+            `,
+            options: [4, { MemberExpression: 2 }]
+        },
+        {
+            code: unIndent`
+                foo
+                    [bar](function() {
+                        baz
+                    })
+            `
+        },
+        {
+            code: unIndent`
+                foo.
+                    bar.
+                    baz
+            `
+        },
+        {
+            code: unIndent`
+                foo
+                    .bar(function() {
+                        baz
+                    })
+            `,
+            options: [4, { MemberExpression: "off" }]
+        },
+        {
+            code: unIndent`
+                foo
+                                .bar(function() {
+                                    baz
+                                })
+            `,
+            options: [4, { MemberExpression: "off" }]
+        },
+        {
+            code: unIndent`
+                foo
+                                [bar](function() {
+                                    baz
+                                })
+            `,
+            options: [4, { MemberExpression: "off" }]
+        },
+        {
+            code: unIndent`
+                  foo.
+                          bar.
+                                      baz
+            `,
+            options: [4, { MemberExpression: "off" }]
+        },
+        {
+            code: unIndent`
+                  foo
+                      [
+                          bar
+                      ]
+                      .baz(function() {
+                          quz();
+                      })
+            `
+        },
+        {
+            code: unIndent`
+                  [
+                      foo
+                  ][
+                      "map"](function() {
+                      qux();
+                  })
+            `
+        },
+        {
+            code: unIndent`
+                (
+                    a.b(function() {
+                        c;
+                    })
+                )
+            `
+        },
+        {
+            code: unIndent`
+                (
+                    foo
+                ).bar(function() {
+                    baz();
+                })
+            `
         }
     ],
 
@@ -6467,6 +6573,19 @@ ruleTester.run("indent", rule, {
                 );
             `,
             errors: expectedErrors([[4, 4, 8, "Identifier"], [5, 0, 4, "Punctuator"]])
+        },
+        {
+            code: unIndent`
+                foo.
+                  bar.
+                      baz
+            `,
+            output: unIndent`
+                foo.
+                    bar.
+                    baz
+            `,
+            errors: expectedErrors([[2, 4, 2, "Identifier"], [3, 4, 6, "Identifier"]])
         }
     ]
 });

